### PR TITLE
Add Github Actions

### PR DESCRIPTION
**Pull request by serban-nicusor-toptal**
Original PR: https://github.com/stan-dev/rstan/pull/833

This PR will try to add more testing to Rstan. The goal is to first test: 

- macOS, R 4.x
- macOS, R 3.6
- windows, R 4.x
- windows, R 3.6
- ubuntu 20.04, R 4.x
- ubuntu 20.04, R 3.6

Next phase would be to also test for: 

- macOS, R-devel
- windows, R-devel
- ubuntu 20.04, R-devel

This is a WIP

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Rok Češnovar, Univ. of Ljubljana

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: GPLv3 (http://opensource.org/licenses/GPL-3.0)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by serban-nicusor-toptal** [comment]
Hey @rok-cesnovar I've created this PR with correct permissions, you are free to edit/commit.
Let me know if I can help with anything.

**Comment by rok-cesnovar** [comment]
Thanks @serban-nicusor-toptal!

I think you just need to copy-paste the file on a new branch that you create? This should already run tests once a PR is open.

**Comment by serban-nicusor-toptal** [comment]
It already started in this PR [here](https://travis-ci.org/github/stan-dev/rstan/builds/720435602?utm_source=github_status&utm_medium=notification)
I've used [add-github-actions-1](https://github.com/stan-dev/rstan/tree/add-github-actions-1) branch

**Comment by rok-cesnovar** [comment]
Ah sorry, I am an idiot. These PRs are done to develop. It should be develop in line 4 and 7. Can you fix those?

**Comment by serban-nicusor-toptal** [comment]
Happens don't worry, I've modified the branch.

**Comment by serban-nicusor-toptal** [comment]
Seems to error out:
```
##[error]Error: File `DESCRIPTION' does not exist
Execution halted
##[error]Process completed with exit code 1.
```

**Comment by rok-cesnovar** [comment]
> Seems to error out:

This one happened because the path to the DESCRIPTION file was not set up correctly in the script. The rstan R package is actually in the ./rstan/rstan folder.

**Comment by rok-cesnovar** [comment]
Ok, these errors that are left now are all non-CI related and need to be investigated further.

**Comment by SteveBronder** [comment]
For MacOS think it needs V8 installed

```
##[error]Package required but not available: ‘V8’
```

Do I have permissions here? I can play around with this tmrw if so

**Comment by rok-cesnovar** [comment]
> Do I have permissions here? I can play around with this tmrw if so

Given that you have permissions to create a branch on this repo (at least you did here https://github.com/stan-dev/rstan/pull/823), you should be able to.

> For MacOS think it needs V8 installed

Yeah, R-devel needs to build it from source (its not build yet for R-devel). In hindsight, we probably dont need to run these for R-devel right now. That can be the next step once we set up testing for the released R version and the last R3.6 release.

**Comment by rok-cesnovar** [comment]
Once we get this green we should add that a successful test also produces binaries for Win/Macs.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        env:
+          _R_CHECK_CRAN_INCOMING_: false
+        run: |
+          if (R.version$major < 4 && isTRUE(.Platform$OS.type == "windows")) {
+            dotR <- file.path(Sys.getenv("HOME"), ".R")
+            if (!file.exists(dotR)) dir.create(dotR)
+            path_makevars <- ifelse(.Platform$OS.type == "windows", "Makevars.win", "Makevars")
+            M <- file.path(dotR, path_makevars)
+            if (!file.exists(M)) file.create(M)
+            write("CXX14=g++ -std=c++1y", file = path_makevars)
```
Comment (rok-cesnovar) [inline .github/workflows/main.yml:90]: I tried the above to get rid of the CXX14 error. I am probably doing something wrong , not sure.

**Review by rok-cesnovar — COMMENTED** [review]
So the current state is: 
- Linux works fine for all versions (R-devel, R4, R3.6)
- Windows 3.6: error about CXX14 (see comment)
- Windows 4.0: error when running examples (this one seems legit), see below for error. I am not exactly sure on what the problem is there.
- Mac 4.0: R CMD CHECK gets stuck at `checking serialization versions ...`. Not completely sure whether the problem is this step or the next one which is `* checking whether package 'rstan' can be installed`. I am guessing its the latter.

Windows 4 error:
```
2020-08-24T09:45:34.0579248Z * checking files in 'vignettes' ... SKIPPED
2020-08-24T09:45:34.0592238Z 
2020-08-24T09:45:34.5871843Z * checking examples ...
2020-08-24T09:45:34.5880435Z 
2020-08-24T09:45:38.1514930Z ** running examples for arch 'i386' ... ERROR
2020-08-24T09:45:38.1518390Z 
2020-08-24T09:45:38.1519173Z Running examples in 'rstan-Ex.R' failed
2020-08-24T09:45:38.1519374Z 
2020-08-24T09:45:38.1519789Z The error most likely occurred in:
2020-08-24T09:45:38.1520710Z 
2020-08-24T09:45:38.1521502Z 
2020-08-24T09:45:38.1521978Z 
2020-08-24T09:45:38.1523964Z > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
2020-08-24T09:45:38.1524257Z 
2020-08-24T09:45:38.1524568Z > ### Name: monitor
2020-08-24T09:45:38.1524780Z 
2020-08-24T09:45:38.1525177Z > ### Title: Compute summaries of MCMC draws and monitor convergence
2020-08-24T09:45:38.1525420Z 
2020-08-24T09:45:38.1525785Z > ### Aliases: monitor print.simsummary [.simsummary
2020-08-24T09:45:38.1525970Z 
2020-08-24T09:45:38.1526228Z > 
2020-08-24T09:45:38.1526401Z 
2020-08-24T09:45:38.1526689Z > ### ** Examples
2020-08-24T09:45:38.1526963Z 
2020-08-24T09:45:38.1527226Z > 
2020-08-24T09:45:38.1527396Z 
2020-08-24T09:45:38.1527767Z > csvfiles <- dir(system.file('misc', package = 'rstan'),
2020-08-24T09:45:38.1528007Z 
2020-08-24T09:45:38.1528399Z +                 pattern = 'rstan_doc_ex_[0-9].csv', full.names = TRUE)
2020-08-24T09:45:38.1528589Z 
2020-08-24T09:45:38.1528910Z > fit <- read_stan_csv(csvfiles)
2020-08-24T09:45:38.1529134Z 
2020-08-24T09:45:38.1529686Z Error in row.buffer[buffer.pointer, ] <- scan(con, nlines = 1, sep = ",",  : 
2020-08-24T09:45:38.1529894Z 
2020-08-24T09:45:38.1530212Z   replacement has length zero
2020-08-24T09:45:38.1530593Z 
2020-08-24T09:45:38.1530904Z Calls: read_stan_csv
2020-08-24T09:45:38.1531083Z 
2020-08-24T09:45:38.1531369Z Execution halted
2020-08-24T09:45:38.1531557Z 
2020-08-24T09:45:41.9834931Z ** running examples for arch 'x64' ... ERROR
2020-08-24T09:45:41.9842175Z 
2020-08-24T09:45:41.9846126Z Running examples in 'rstan-Ex.R' failed
2020-08-24T09:45:41.9846902Z 
2020-08-24T09:45:41.9847752Z The error most likely occurred in:
2020-08-24T09:45:41.9848033Z 
2020-08-24T09:45:41.9848210Z 
2020-08-24T09:45:41.9848429Z 
2020-08-24T09:45:41.9848962Z > base::assign(".ptime", proc.time(), pos = "CheckExEnv")
2020-08-24T09:45:41.9849164Z 
2020-08-24T09:45:41.9849474Z > ### Name: monitor
2020-08-24T09:45:41.9849706Z 
2020-08-24T09:45:41.9850108Z > ### Title: Compute summaries of MCMC draws and monitor convergence
2020-08-24T09:45:41.9850301Z 
2020-08-24T09:45:41.9850670Z > ### Aliases: monitor print.simsummary [.simsummary
2020-08-24T09:45:41.9850899Z 
2020-08-24T09:45:41.9851118Z > 
2020-08-24T09:45:41.9851351Z 
2020-08-24T09:45:41.9851642Z > ### ** Examples
2020-08-24T09:45:41.9852534Z 
2020-08-24T09:45:41.9853495Z > 
2020-08-24T09:45:41.9853963Z 
2020-08-24T09:45:41.9855144Z > csvfiles <- dir(system.file('misc', package = 'rstan'),
2020-08-24T09:45:41.9855368Z 
2020-08-24T09:45:41.9855782Z +                 pattern = 'rstan_doc_ex_[0-9].csv', full.names = TRUE)
2020-08-24T09:45:41.9856027Z 
2020-08-24T09:45:41.9856352Z > fit <- read_stan_csv(csvfiles)
2020-08-24T09:45:41.9856531Z 
2020-08-24T09:45:41.9857435Z Error in row.buffer[buffer.pointer, ] <- scan(con, nlines = 1, sep = ",",  : 
2020-08-24T09:45:41.9857904Z 
2020-08-24T09:45:41.9858228Z   replacement has length zero
2020-08-24T09:45:41.9858426Z 
2020-08-24T09:45:41.9858747Z Calls: read_stan_csv
2020-08-24T09:45:41.9858981Z 
2020-08-24T09:45:41.9859274Z Execution halted
2020-08-24T09:45:41.9859493Z 
2020-08-24T09:45:42.0885954Z * checking for unstated dependencies in 'tests' ... OK
2020-08-24T09:45:42.0886531Z 
2020-08-24T09:45:42.0886911Z * checking tests ...
2020-08-24T09:45:42.0887155Z 
2020-08-24T09:45:42.3048138Z ** running tests for arch 'i386' ...
2020-08-24T09:45:42.3048594Z 
2020-08-24T09:45:45.7106633Z   Running 'doRUnit.R'
2020-08-24T09:45:45.7107553Z 
2020-08-24T09:45:45.7213413Z  OK
2020-08-24T09:45:45.7213931Z 
2020-08-24T09:45:45.9276974Z ** running tests for arch 'x64' ...
2020-08-24T09:45:45.9277449Z 
2020-08-24T09:45:49.2560118Z   Running 'doRUnit.R'
2020-08-24T09:45:49.2560639Z 
2020-08-24T09:45:49.2688225Z  OK
2020-08-24T09:45:49.2696340Z 
2020-08-24T09:45:49.2719907Z * checking for non-standard things in the check directory ... OK
2020-08-24T09:45:49.2728358Z 
2020-08-24T09:45:49.2728930Z * checking for detritus in the temp directory ... OK
2020-08-24T09:45:49.2729125Z 
2020-08-24T09:45:49.2729489Z * DONE
```

**Comment by jgabry** [comment]
Thanks for working on this! 

> * Mac 4.0: R CMD CHECK gets stuck at `checking serialization versions ...`. Not completely sure whether the problem is this step or the next one which is `* checking whether package 'rstan' can be installed`. I am guessing its the latter.

I'm now seeing a different error when I check the Mac results: 

```
Error: <rlib_error in process_get_error_connection(self, private):
 stderr is not a pipe.>
```

I've never seen that error before when running R CMD check. Judging from the stack trace I think the [rcmdcheck](https://github.com/r-lib/rcmdcheck) package is calling the [processx](https://github.com/r-lib/processx) package and it's processx that's throwing this particular error. Those packages actually have the same maintainer, so if necessary we can contact them if we can't figure out what's going on.

**Comment by rok-cesnovar** [comment]
> Error: <rlib_error in process_get_error_connection(self, private):
>  stderr is not a pipe.>

Yeah, that error is printed after it runs for 6 hours stuck on that step and it probably times-out. That would be my guess, it might also be rcmdcheck.

**Comment by SteveBronder** [comment]
> `The error most *likely* occurred in:`

lol'd at most likely, it really can't tell?

Yeah I can take a look at this tmrw. My guess is it's legit and we need to do some check to see if we are on windows or not

**Pull request merged**